### PR TITLE
H liu

### PR DESCRIPTION
**Pull request by hliu26**
Hi Professor Sun, 
The files should have docstrings and error raising/ checking of parameters for each distribution.

### CODE REVIEW

File: symbulate/distributions.py [hunk 1]
```diff
@@ -91,8 +93,16 @@ class Binomial(Distribution):
     """
 
     def __init__(self, n, p):
-        self.n = n
-        self.p = p
+        
+        if n >= 0:
```
Comment (dlsun) [inline symbulate/distributions.py]: n also has to be an integer.

File: symbulate/distributions.py [hunk 2]
```diff
@@ -118,19 +131,38 @@ class Hypergeometric(Distribution):
     """
 
     def __init__(self, n, N0, N1):
-        self.n = n
-        self.N0 = N0
-        self.N1 = N1
         
+        if n > 0:
```
Comment (dlsun) [inline symbulate/distributions.py]: Ditto, check that n, N0, N1 are integers.

File: symbulate/distributions.py [hunk 3]
```diff
@@ -118,19 +131,38 @@ class Hypergeometric(Distribution):
...
+            raise Exception("Number of successes cannot be negative")
+
         params = {
             "M" : N0 + N1,
             "n" : N1,
             "N" : n
             }
+
+        if (N0 + N1) < n:
+            raise Exception("Number of Successes + Failures cannot be less than the sample size N")
```
Comment (dlsun) [inline symbulate/distributions.py]: Replace "sample size N" by "sample size n". (Uppercase N refers to the population size.)

File: symbulate/distributions.py [hunk 4]
```diff
@@ -78,6 +77,9 @@ def __init__(self, p):
         self.xlim = (0, 1) # Bernoulli distributions are not defined for x < 0 and x > 1
  
     def draw(self):
+        """A function that takes no arguments and 
+            takes a single draw from the Bernoulli distribution."""
```
Comment (dlsun) [inline symbulate/distributions.py]: I would word this as "...and returns a single draw..."

Ditto for all of the docstrings below.

File: symbulate/distributions.py [hunk 5]
```diff
@@ -169,8 +208,16 @@ class NegativeBinomial(Distribution):
     """
 
     def __init__(self, r, p):
-        self.r = r
-        self.p = p
+
+        if 0 < r:
```
Comment (dlsun) [inline symbulate/distributions.py]: Check that r is an integer.

File: symbulate/distributions.py [hunk 6]
```diff
@@ -198,9 +248,17 @@ class Pascal(Distribution):
     """
     
     def __init__(self, r, p):
-        self.r = r
-        self.p = p
         
+        if 0 < r:
```
Comment (dlsun) [inline symbulate/distributions.py]: Same as NegativeBinomial.

File: symbulate/distributions.py [hunk 7]
```diff
@@ -270,15 +345,21 @@ def __init__(self, mean=0.0, var=1.0, sd=None):
         if sd is None:
             self.scale = np.sqrt(var)
         else:
-            self.scale = sd
-        
+            if sd > 0:
```
Comment (dlsun) [inline symbulate/distributions.py]: In that case, you should also check that var > 0.

File: symbulate/distributions.py [hunk 8]
```diff
@@ -380,15 +492,33 @@ def __init__(self, mean, cov):
             raise Exception("The dimension of the mean vector" +
                             "is not compatible with the dimensions" +
                             "of the covariance matrix.")
-        self.mean = mean
-        self.cov = cov
+        
+        if len(mean) >= 2:
```
Comment (dlsun) [inline symbulate/distributions.py]: I like your thinking, but I'm not sure I agree with this. You should be able to simulate a MultivariateNormal where the number of dimensions n=1. Of course, this is equivalent to just simulating a Normal.

File: symbulate/probability_space.py [hunk 9]
```diff
@@ -138,6 +138,10 @@ def __init__(self, box, size=None, replace=True, probs=None, order_matters=True)
         self.order_matters = order_matters
 
     def draw(self):
+        """A function, that takes no arguments, explaining how to draw one
```
Comment (dlsun) [inline symbulate/probability_space.py]: This docstring is not helpful to the user. A user will want to know what .draw() does for a BoxModel. From their perspective, what they want to know is what happens when they call .draw() for a BoxModel. They won't be implementing the .draw() on their own.

File: symbulate/random_variables.py [hunk 10]
```diff
@@ -11,9 +11,22 @@ def __init__(self, probSpace, fun=lambda x: x):
         self.fun = fun
 
     def draw(self):
+        """A function, that takes no arguments, explaining how to draw one
```
Comment (dlsun) [inline symbulate/random_variables.py]: As above, the user wants to know what will happen when they call .draw() on the RV. What arguments does this function take and what does it return?

File: symbulate/random_variables.py [hunk 11]
```diff
@@ -23,6 +36,13 @@ def check_same_probSpace(self, other):
             self.probSpace.check_same(other.probSpace)
 
     def apply(self, function):
+        """
+        Args:
+            function: function being used on results (addition, subtractions, etc.)
```
Comment (dlsun) [inline symbulate/random_variables.py]: Suggestion: replace by 

function: function to apply to the random variable (e.g., log, sqrt, exp)

File: symbulate/random_variables.py [hunk 12]
```diff
@@ -233,6 +253,11 @@ def __init__(self, random_variable, condition_event):
                          random_variable.fun)
         
     def draw(self):
+        """
+            A function, that takes no arguments, explaining how to draw one
```
Comment (dlsun) [inline symbulate/random_variables.py]: Ditto as above. Please make this docstring more helpful to an end user.

File: symbulate/random_variables.py [hunk 13]
```diff
@@ -11,9 +11,22 @@ def __init__(self, probSpace, fun=lambda x: x):
         self.fun = fun
```
Comment (dlsun) [inline symbulate/random_variables.py]: This constructor needs a docstring as well.

File: symbulate/distributions.py (hunk 7 above)
Comment (hliu26) [inline symbulate/distributions.py]: Hi Professor, I'm slightly confused on what you mean or where the check should be placed?
Should I check in the else: if var > 0: then sd = np.sqrt(var) and else: var can't be less than 0?

File: symbulate/random_variables.py [hunk 14]
```diff
@@ -254,10 +259,16 @@ def __init__(self, random_variable, condition_event):
         
     def draw(self):
         """
-            A function, that takes no arguments, explaining how to draw one 
-        outcome from the Probability Space described by the random variable.
+        A function that takes no arguments and returns a single set 
+            of values from the probability spaces defined in the conditional
```
Comment (dlsun) [inline symbulate/random_variables.py]: Again, you're not returning values from the probability space. You're returning values of the random variable. Please rewrite this.

File: symbulate/random_variables.py [hunk 15]
```diff
@@ -254,10 +259,16 @@ def __init__(self, random_variable, condition_event):
         
     def draw(self):
         """
-            A function, that takes no arguments, explaining how to draw one 
-        outcome from the Probability Space described by the random variable.
+        A function that takes no arguments and returns a single set 
+            of values from the probability spaces defined in the conditional 
+            random variable that meet the conditional event. 
+        The number of values returned in the set depend on the number of
```
Comment (dlsun) [inline symbulate/random_variables.py]: This comment is confusing and should be removed. We're not defining multiple probability spaces here; we're defining multiple random variables.

**Comment by dlsun** [comment]
A few more small things, and then I think it's ready to be merged.

**Pull request merged**